src/gtk/window.cpp

```diff
@@ -2828,6 +2893,661 @@ static gboolean source_dispatch(GSource*, GSourceFunc, void*)
+
+        m_rotate_gesture = gtk_gesture_rotate_new(widget);
+
+        gtk_event_controller_set_propagation_phase (GTK_EVENT_CONTROLLER(m_rotate_gesture), GTK_PHASE_TARGET);
+
+        g_signal_connect (m_rotate_gesture, "begin",
+                          G_CALLBACK(rotate_gesture_begin_callback), win);
+        g_signal_connect (m_rotate_gesture, "angle-changed",
+                          G_CALLBACK(rotate_gesture_callback), win);
+        g_signal_connect (m_rotate_gesture, "end",
+                          G_CALLBACK(rotate_gesture_end_callback), win);
+        g_signal_connect (m_rotate_gesture, "cancel",
+                          G_CALLBACK(rotate_gesture_end_callback), win);
+    }
+    else
+    {
+        m_rotate_gesture = NULL;
+    }
+
+    if ( eventsMask & wxTOUCH_PRESS_GESTURES )
+    {
+        eventsMask &= ~wxTOUCH_PRESS_GESTURES;
+
+        m_long_press_gesture = gtk_gesture_long_press_new(widget);
+
+        // "pressed" signal is also generated when left mouse is down for some minimum duration of time.
+        // This can be disable by calling gtk_gesture_single_set_touch_only(GTK_GESTURE_SINGLE(m_long_press_gesture), TRUE)
+        // which will allow "pressed" signal only for Touch events.
+
+        gtk_event_controller_set_propagation_phase(GTK_EVENT_CONTROLLER(m_long_press_gesture), GTK_PHASE_TARGET);
+
+        g_signal_connect (m_long_press_gesture, "pressed",
+                          G_CALLBACK(long_press_gesture_callback), win);
+    }
+    else
+    {
+        m_long_press_gesture = NULL;
+    }
+
+    wxASSERT_MSG( eventsMask == 0, "Unknown touch event mask bit specified" );
+
+    g_signal_connect (widget, "touch-event",
+                      G_CALLBACK(touch_callback), win);
+}
+
+wxWindowGesturesData::~wxWindowGesturesData()
+{
+    if ( m_vertical_pan_gesture )
+        g_object_unref(m_vertical_pan_gesture);
+    if ( m_horizontal_pan_gesture )
+        g_object_unref(m_horizontal_pan_gesture);
+    if ( m_zoom_gesture )
+        g_object_unref(m_zoom_gesture);
+    if ( m_rotate_gesture )
+        g_object_unref(m_rotate_gesture);
+    if ( m_long_press_gesture )
+        g_object_unref(m_long_press_gesture);
+}
+
+#endif // wxGTK_HAS_GESTURES_SUPPORT
+
+// This method must be always defined for GTK+ 3 as it's declared in the
+// header, where we can't (easily) test for wxGTK_HAS_GESTURES_SUPPORT.
+#ifdef __WXGTK3__
+
+bool wxWindowGTK::EnableTouchEvents(int eventsMask)
+{
+#ifdef wxGTK_HAS_GESTURES_SUPPORT
+    // Check if gestures support is also available during run-time.
+    if ( gtk_check_version(3, 14, 0) == NULL )
+    {
+        if ( eventsMask == wxTOUCH_NONE )
+        {
+            wxWindowGestures::EraseForObject(this);
+        }
+        else
+        {
+            wxWindowGestures::StoreForObject
+            (
+                this,
+                new wxWindowGesturesData(this, GetConnectWidget(), eventsMask)
+            );
+        }
```

Comment:
It is very likely, that we will see more Options, which might not be implemented in all backend.  The current code accepts not handled bits by returning true.
If a programmer wants to enable all events specified in the mask, this behaviour is bad. Another programmer might accept, that some events can't be enabled.

What if EnableTouchEvents returns the list of enabled events instead of true/false? This way it would work for both use cases.

Reply:
To be honest, I think that the return value will mostly be ignored anyhow (what can you do if it fails?). But returning the mask of existing events could be useful... I really don't know if I'm going to have time to work on this myself though, but then I don't think this is critical to do before the merge.